### PR TITLE
ggml : add ggml_soft_max_ext

### PR DESCRIPTION
**Pull request by ggerganov**
Fused op:

```
soft_max_ext(x) = soft_max(x*scale + mask)
```

Should improve PP performance for large contexts ( > 2k)

Also:
- ~multi-thread `ggml_soft_max` on CPU~
- CUDA and Metal soft max kernels use up to 1024 threads (based on dim 0) instead of always 32 and perform warp-based reduction
- Metal rms_norm kernel changed to perform warp-based reduction as well

## Perf CUDA V100

```bash
LLAMA_CUBLAS=1 make -j llama-bench && ./llama-bench -m ./models/openllama-7b-v2/ggml-model-q4_0.gguf -p 512,1024,4096,8192 -ngl 99
```

ggml_init_cublas: GGML_CUDA_FORCE_MMQ:   no
ggml_init_cublas: CUDA_USE_TENSOR_CORES: yes
ggml_init_cublas: found 1 CUDA devices:
  Device 0: Tesla V100-PCIE-16GB, compute capability 7.0
| model                   |       size | backend    | ngl | test       |    master    t/s |    PR t/s |       speedup |
| ----------------------- | ---------: | ---------- | --: | ---------- | ---------------: | ---------------: | -----: |
| llama 7B Q4_0           |   3.56 GiB | CUDA       |  99 | pp 512     |  2661.62 ± 71.30 |  2659.81 ± 65.20 | 0.999 |
| llama 7B Q4_0           |   3.56 GiB | CUDA       |  99 | pp 1024    |  2515.25 ± 39.75 |  2577.93 ± 46.42 | 1.025 |
| llama 7B Q4_0           |   3.56 GiB | CUDA       |  99 | pp 4096    |  1793.41 ± 24.16 |  2106.14 ± 29.02 | 1.174 |
| llama 7B Q4_0           |   3.56 GiB | CUDA       |  99 | pp 8192    |  1297.36 ± 12.95 |  1552.12 ± 18.71 | 1.196 |
| llama 7B Q4_0           |   3.56 GiB | CUDA       |  99 | tg 128     |    112.52 ± 0.64 |    114.98 ± 0.58 | 1.022 |

```bash
LLAMA_CUBLAS=1 make -j batched-bench && ./batched-bench ./models/openllama-7b-v2/ggml-model-q4_0.gguf 10880 0 99 0 2048 128 1,2,4,5
```

`master`
|    PP |     TG |    B |   N_KV |   T_PP s | S_PP t/s |   T_TG s | S_TG t/s |      T s |    S t/s |
|-------|--------|------|--------|----------|----------|----------|----------|----------|----------|
|  2048 |    128 |    1 |   2176 |    0.982 |  2085.81 |    1.703 |    75.16 |    2.685 |   810.45 |
|  2048 |    128 |    2 |   4352 |    2.329 |  1758.55 |    2.587 |    98.94 |    4.917 |   885.17 |
|  2048 |    128 |    4 |   8704 |    6.415 |  1276.99 |    3.737 |   137.02 |   10.152 |   857.38 |
|  2048 |    128 |    5 |  10880 |    9.177 |  1115.80 |    5.052 |   126.67 |   14.230 |   764.60 |

PPL = 7.0773 +/- 0.04208

`PR`
|    PP |     TG |    B |   N_KV |   T_PP s | S_PP t/s |   T_TG s | S_TG t/s |      T s |    S t/s |
|-------|--------|------|--------|----------|----------|----------|----------|----------|----------|
|  2048 |    128 |    1 |   2176 |    0.908 |  2255.56 |    1.645 |    77.80 |    2.553 |   852.25 |
|  2048 |    128 |    2 |   4352 |    1.994 |  2054.42 |    2.509 |   102.04 |    4.502 |   966.57 |
|  2048 |    128 |    4 |   8704 |    5.375 |  1524.08 |    3.553 |   144.12 |    8.928 |   974.95 |
|  2048 |    128 |    5 |  10880 |    7.669 |  1335.30 |    4.732 |   135.26 |   12.400 |   877.40 |

PPL = 7.0768 +/- 0.04207

## Perf M2 Ultra

```bash
make -j llama-bench && ./llama-bench -m ./models/llama-7b-v2/ggml-model-q4_0.gguf -p 512,1024,4096,8192 -ngl 99
```

| model                   |       size | backend    | ngl | test       |     master   t/s |    PR        t/s | speedup |
| ----------------------- | ---------: | ---------- | --: | ---------- | ---------------: | ---------------: | ------: |
| llama 7B Q4_0           |   3.56 GiB | Metal      |  99 | pp 512     |   1238.73 ± 0.94 |   1275.95 ± 1.05 |    1.03 |
| llama 7B Q4_0           |   3.56 GiB | Metal      |  99 | pp 1024    |   1206.55 ± 0.34 |   1260.08 ± 0.96 |    1.04 |
| llama 7B Q4_0           |   3.56 GiB | Metal      |  99 | pp 4096    |   1080.81 ± 0.30 |   1154.05 ± 0.30 |    1.07 |
| llama 7B Q4_0           |   3.56 GiB | Metal      |  99 | pp 8192    |    933.17 ± 0.32 |   1023.16 ± 1.19 |    1.10 |
| llama 7B Q4_0           |   3.56 GiB | Metal      |  99 | tg 128     |     94.40 ± 0.04 |     96.23 ± 0.06 |    1.02 |

```bash
make -j batched-bench && ./batched-bench ./models/llama-7b-v2/ggml-model-q4_0.gguf 10880 0 99 0 2048 128 1,2,4,5
```

`master`
|    PP |     TG |    B |   N_KV |   T_PP s | S_PP t/s |   T_TG s | S_TG t/s |      T s |    S t/s |
|-------|--------|------|--------|----------|----------|----------|----------|----------|----------|
|  2048 |    128 |    1 |   2176 |    1.770 |  1157.14 |    1.785 |    71.71 |    3.555 |   612.10 |
|  2048 |    128 |    2 |   4352 |    3.793 |  1079.79 |    7.648 |    33.47 |   11.442 |   380.36 |
|  2048 |    128 |    4 |   8704 |    8.774 |   933.72 |    9.399 |    54.48 |   18.172 |   478.98 |
|  2048 |    128 |    5 |  10880 |   11.817 |   866.53 |   10.285 |    62.22 |   22.103 |   492.25 |

PPL = 6.1212 +/- 0.03511

`PR`
|    PP |     TG |    B |   N_KV |   T_PP s | S_PP t/s |   T_TG s | S_TG t/s |      T s |    S t/s |
|-------|--------|------|--------|----------|----------|----------|----------|----------|----------|
|  2048 |    128 |    1 |   2176 |    1.677 |  1221.20 |    1.741 |    73.54 |    3.418 |   636.70 |
|  2048 |    128 |    2 |   4352 |    3.512 |  1166.13 |    7.555 |    33.89 |   11.067 |   393.24 |
|  2048 |    128 |    4 |   8704 |    7.786 |  1052.16 |    9.099 |    56.27 |   16.885 |   515.49 |
|  2048 |    128 |    5 |  10880 |   10.243 |   999.71 |   10.004 |    63.98 |   20.247 |   537.38 |

PPL = 6.1212 +/- 0.03511

```bash
make -j llama-bench && ./llama-bench -m ./models/llama-7b-v2/ggml-model-q4_0.gguf -p 512 -n 128,256,512 -ngl 99
```

`master`
| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | pp 512     |   1237.37 ± 1.50 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 128     |     94.49 ± 0.03 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 256     |     93.79 ± 0.04 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 512     |     92.37 ± 0.18 |

build: 1f5cd832 (1577)

`PR`
| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | pp 512     |   1269.86 ± 1.40 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 128     |     98.14 ± 0.02 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 256     |     97.36 ± 0.06 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 512     |     95.78 ± 0.05 |

## Perf M1 Pro

```bash
make -j llama-bench && ./llama-bench -m ./models/llama-7b-v2/ggml-model-q4_0.gguf -p 512 -n 128,256,512 -ngl 99
```

`master`
| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | pp 512     |    264.95 ± 0.92 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 128     |     36.52 ± 0.13 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 256     |     36.21 ± 0.01 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 512     |     35.54 ± 0.01 |

build: 1f5cd832 (1577)

`PR`
| model                          |       size |     params | backend    | ngl | test       |              t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | ---------- | ---------------: |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | pp 512     |    273.44 ± 0.90 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 128     |     37.18 ± 0.03 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 256     |     36.80 ± 0.03 |
| llama 7B mostly Q4_0           |   3.56 GiB |     6.74 B | Metal      |  99 | tg 512     |     36.20 ± 0.01 |

build: c4db5923 (1589)

### CODE REVIEW

**Comment by ggerganov** [comment]
I think the CPU performance is degraded atm (did some quick tests earlier, haven't looked deeper yet). But I suspect it's something to do with the threading.

File: ggml-cuda.cu [hunk 1]
```diff
@@ -6846,14 +6849,18 @@ inline void ggml_cuda_op_soft_max(
     GGML_ASSERT(src0->type == GGML_TYPE_F32);
     GGML_ASSERT( dst->type == GGML_TYPE_F32);
 
+    GGML_ASSERT(!src1 || src1->type == GGML_TYPE_F32); // src1 contains mask and it is optional
+
     const int64_t ne00 = src0->ne[0];
-    const int64_t nrows = ggml_nrows(src0);
+    const int64_t nrows_x = ggml_nrows(src0);
+    const int64_t nrows_y = src1 ? ggml_nrows(src1) : 0;
```
Comment (ggerganov) [inline ggml-cuda.cu:6856]:
```suggestion
    const int64_t nrows_y = src1 ? ggml_nrows(src1) : 1;
```

**Review by ggerganov — COMMENTED** [review]
Should be ready to merge. Improved performance with Metal and CUDA. On CPU - not much difference

File: ggml-cuda.cu [hunk 2]
```diff
@@ -4717,45 +4718,62 @@ static __global__ void diag_mask_inf_f32(const float * x, float * dst, const int
...
-    const int tid = threadIdx.y;
+// TODO: maybe can be improved with some warp-based primitives
+static __global__ void soft_max_f32(const float * x, const float * y, float * dst, const int ncols, const int nrows_y, const float scale) {
+    const int tid  = threadIdx.x;
+    const int rowx = blockIdx.x;
+    const int rowy = rowx % nrows_y; // broadcast the mask (y) in the row dimension
+
+    const int block_size = blockDim.x;
+
+    __shared__ float buf[CUDA_SOFT_MAX_BLOCK_SIZE];
```
[the remark this answers is not in the document]
Comment (slaren) [reply ggml-cuda.cu:4729]: Block wide reductions are a lot more expensive than warp reductions, so most of the kernels only distribute operations that need a reduction to a warp. A block reduction is usually done by doing a reduction in every warp and storing the result of each warp to shared memory, and then doing a reduction of the shared memory.

[the remark this answers is not in the document]
Comment (ggerganov) [reply ggml-cuda.cu]: Updated to use warp-based block reduction with max of 32 warps. Results improved, more significantly for the largest contexts.

[the remark this answers is not in the document]
Comment (ggerganov) [reply ggml-cuda.cu]: I'll try to do similar optimization for the Metal kernel

**Comment by slaren** [comment]
Much larger speedup now, but it is probably broken, perplexity returns only nans.

| model                          | backend    | ngl | test       |  master 1f5cd832 | PR 6b86bcff | speedup |
| ------------------------------ | ---------- | --: | ---------- | ---------------: | ----------: | ------: |
| llama 7B mostly Q4_0           | CUDA       |  99 | pp 512     |  3761.61 ± 76.47 | 3848.99 ± 68.95 | 1.023 |
| llama 7B mostly Q4_0           | CUDA       |  99 | pp 1024    |  3565.39 ± 37.27 | 3646.32 ± 39.29 | 1.022 |
| llama 7B mostly Q4_0           | CUDA       |  99 | pp 4096    |  2523.32 ± 12.70 | 3039.41 ± 23.85 | 1.204 |
| llama 7B mostly Q4_0           | CUDA       |  99 | pp 8192    |   1776.08 ± 8.35 | 2502.62 ± 15.85 | 1.409 |
| llama 7B mostly Q4_0           | CUDA       |  99 | pp 16384   |   1124.91 ± 6.93 | 1745.35 ± 16.31 | 1.551 |
| llama 7B mostly Q4_0           | CUDA       |  99 | tg 128     |    124.38 ± 0.81 |   127.22 ± 1.15 | 1.022 |

**Comment by ggerganov** [comment]
Thanks for checking - the shared arrays were not initialized so the warps were reducing random data. Should be fixed now

**Pull request merged**